### PR TITLE
Fix disable_sstable_write synchronization with on_compaction_completion

### LINKED ISSUE #4622: disrupt_nodetool_refresh fails with error sstables::malformed_sstable_exception on 3.1
it was observed during longevity 4h (multiple times) for scylla 3.1.

latest seen on [job from June 26th](https://jenkins.scylladb.com/view/scylla-3.1/job/scylla-3.1/job/longevity/job/longevity-100gb-4h/25/)

tar ball with files to be loaded to the DB attached.

the nemesis executes those steps:

1. download the file from S3 [keyspace1.standard1.tar.gz](https://github.com/scylladb/scylla/files/3342644/keyspace1.standard1.tar.gz)
2. check if keyspace1.standard1 exists
3. find the latest "data" file on /var/lib/scylla/data/keyspace1/
4. untar file to data folder
5. run command "nodetool --host localhost refresh -- keyspace1 standard1"
6. check if content was loaded to the DB

the disruption fails on step 5 with following error:
```
2019-06-26 07:55:41,830 remote           L0682 INFO | RemoteCmdRunner [centos@172.30.0.71]: nodetool: Scylla API server HTTP POST to URL '/storage_service/sstables/keyspace1' failed: Failed to load new sstables: sstables::malformed_sstable_exception (/var/lib/scylla/data/keyspace1/standard1-4ec0213097e511e98877000000000000/mc-142-big-TOC.txt: file not found)
2019-06-26 07:55:41,831 remote           L0682 INFO | RemoteCmdRunner [centos@172.30.0.71]: See 'nodetool help' or 'nodetool help <command>'.
2019-06-26 07:55:41,870 remote           L0140 INFO | RemoteCmdRunner [centos@172.30.0.71]: Command "nodetool --host localhost refresh -- keyspace1 standard1" finished with status 1
2019-06-26 07:55:41,870 nemesis          L0177 ERROR| sdcm.nemesis.ChaosMonkey: nodetool command 'nodetool --host localhost refresh -- keyspace1 standard1' failed on node Node longevity-100gb-4h-3-1-db-node-837e847a-3 [3.89.215.22 | 172.30.0.71] (seed: False): Command exited with status 1.
=== stdout ===
nodetool: Scylla API server HTTP POST to URL '/storage_service/sstables/keyspace1' failed: Failed to load new sstables: sstables::malformed_sstable_exception (/var/lib/scylla/data/keyspace1/standard1-4ec0213097e511e98877000000000000/mc-142-big-TOC.txt: file not found)
See 'nodetool help' or 'nodetool help <command>'.
```

### PR DESCRIPTION
**Pull request by bhalevy**
`disable_sstable_write` needs to acquire `_sstable_deletion_sem` to properly synchronize with background deletions done by `on_compaction_completion` to ensure no sstables will be created or deleted during `reshuffle_sstables` after `storage_service::load_new_sstables` disables sstable writes.

Fixes #4622

Test: unit(dev), nodetool_additional_test.py migration_test.py

### CODE REVIEW

**Comment by raphaelsc** [comment]
other than my concern above, it looks good to me

**Comment by bhalevy** [comment]
> wouldn't this bloat the log with tons of message when LCS is used for example?

I think the overhead is worth it and we're logging anyway when compacting each sstable so it should be comparable.

**Comment by glommer** [comment]
One compaction can involve upwards to 32 SSTables, and have it happening from all shards.

Keep in mind that there is a plan to remove even the compaction messages we have today. Avi even merged a patch for that at some point and I begged him to revert it, because the compaction table as we have to day does not have the full info needed for that -> but when it does, the plan is still to remove even the existing messages.

I am totally against moving this message to info.

**Comment by bhalevy** [comment]
OK, I'll drop this patch then.

**Comment by bhalevy** [comment]
@glommer
In v2 ([6dad9ba](https://github.com/bhalevy/scylla/commits/6dad9baa1c0e4634f69d698612dd3d5dd7133f1d)):
- dropped `sstable: remove_by_toc_name: log info message`
- rebased and retested unit(dev)

File: table.cc [hunk 1]
```diff
@@ -1958,6 +1958,7 @@ future<int64_t>
 table::disable_sstable_write() {
     _sstable_writes_disabled_at = std::chrono::steady_clock::now();
     return _sstables_lock.write_lock().then([this] {
+      return _sstable_deletion_sem.wait().then([this] {
         if (_sstables->all()->empty()) {
```
Comment (avikivity) [inline table.cc]: Please document the lock order in a follow up.

Perhaps we should fold _sstable_deletion_sem into _sstables_lock.

**Comment by avikivity** [comment]
@glommer are you done with your review?

**Comment by glommer** [comment]
Yes, the only thing unreasonable for me was the info message.

**Comment by bhalevy** [comment]
@avikivity can you please merge this?

File: sstables/sstables.cc [hunk 2]
```diff
@@ -2907,7 +2907,7 @@ remove_by_toc_name(sstring sstable_toc_name, const io_error_handler& error_handl
         auto new_toc_name = prefix + sstable_version_constants::TEMPORARY_TOC_SUFFIX;
         sstring dir;
 
-        sstlog.debug("Removing by TOC name: {}", sstable_toc_name);
+        sstlog.info("Removing SSTable {}", sstable_toc_name);
```
[the remark this answers is not in the document]
Comment (slivne) [reply sstables/sstables.cc]: @bhalevy did you address this ?

[the remark this answers is not in the document]
Comment (bhalevy) [reply sstables/sstables.cc]: Yes. As I mentioned:

> In v2 ([6dad9ba](https://github.com/bhalevy/scylla/commits/6dad9baa1c0e4634f69d698612dd3d5dd7133f1d)):
> 
> * dropped `sstable: remove_by_toc_name: log info message`
> * rebased and retested unit(dev)

**Comment by slivne** [comment]
@avikivity please merge we need this for 3.1

File: table.cc (hunk 1 above)
Comment (bhalevy) [reply table.cc]: documented in 0e4567c8815e4bc7f6acc72aee88194e75a2107f

Regarding folding into `_sstables_lock` , they are currently taken at different times in the table::compact_sstables -> on_compaction_completion path, and I'm not sure about the table::snapshot path.  I think that keeping them separate makes sense.

**Comment by bhalevy** [comment]
@avikivity please backport to 3.1 and 3.0

**Comment by avikivity** [comment]
Backported to 3.1.

**Comment by avikivity** [comment]
@bhalevy please prepare a 3.0 backport

**Comment by bhalevy** [comment]
@avikivity 3.0 backport here: https://github.com/bhalevy/scylla/commits/fix-disable-sstable-write-for-3.0
based on branch-3.0 d77aaada86885b1a08b90a9a6cda3379cd0735a9

**Comment by avikivity** [comment]
thanks, queued as a198db31dc754fcfc1170c7be8acba8e01955a13

**Pull request merged**